### ISSUE TITLE
Faulty core fixes: Comma incorrectly added to function call in array declaration

### ISSUE THREAD

**Issue by pento** [WordPress-Coding-Standards/WordPress-Coding-Standards#998]
When given the following input:

```php
<?php

array(
	'foo' => bar(1, 29));
```

`phpcbf` gives me this output:

```php
<?php

array(
	'foo' => bar( 1, 29, ),
);
```

The extra comma in the `bar()` call is being added by the `WordPress.Arrays.ArrayDeclaration.NoComma` rule.

In Core, this can be seen when running against `tests/phpunit/tests/pomo/po.php`.

**Comment by jrfnl** [comment]
I've just ran an isolated test case to try and reproduce the issue and I'm getting different results - though still very wrong - when running that particular snippet against `WordPress-Core`. I'm seeing getting fixed as:
```php
array(
	'foo' => bar( 1, 29
	);
```

=> Closing parenthesis for the `bar()` function call being removed

All the same, it does mean that based on this snippet, I cannot reproduce the issue which was reported.

**Comment by jrfnl** [comment]
Running against `tests/phpunit/tests/pomo/po.php`, I can against see several more cases where, when there are functions or other arrays nested within arrays, one of the closing parenthesis disappears, but so far I have not been able to find one where the extra comma gets added.

**Comment by pento** [comment]
Oh, I found the difference. I'm excluding the `WordPress.Arrays.ArrayIndentation`rule - adding that back in gives me the same result as you.

**Comment by jrfnl** [comment]
@pento In that case, let's change the title and issue description so this issue addresses the "disappearing array end brace" which is in fact a real issue.

FYI: I'm working on a fix as we speak.

**Comment by jrfnl** [comment]
PR #1003 will fix the "case of the disappearing array close brace". However, with that issue fixed, I can now reproduce the extra comma being added....

Will look into that as well.

**Comment by jrfnl** [comment]
Ok, so I've isolated this all the way down.

The minimum code snippet needed with the least interference of other sniffs is this:
```php
$a = array(
	'foo' => bar( 1, 29 ));
```

The issue is caused by the `ArrayDeclaration` sniff. This sniff is a copy of the `Squiz` version with some parts disabled by commenting them out.

The Squiz version works fine, but this is more by accident then by design in this case.
The reason why the Squiz version does not run into this problem is because the `Squiz.Arrays.ArrayDeclaration.CloseBraceNotAligned` fixer kicks in and adjusts the same token causing the "comma" correction to be moved to the next loop where it works fine.
WPCS has commented out the `CloseBraceNotAligned` fixer which is why we're now seeing this issue.

I'm working on creating a stand-alone sniff to deal with comma's after array items to fix this.

------
Additional info for anyone coming across this issue and wondering why WPCS uses a copy of the Squiz sniff with commented out code:

This is/was necessary as:
a) This was the only sniff available with any sort of array format checking and did most of what was needed for WPCS
b) The sniff has not been set up that well, as when it throws an error, it continues to the next loop instead of *also* doing the other array checks. Excluding sniff codes would not help in this case as the `continue` would still be called, which is why bits are commented out instead.

There is a long standing issue open upstream about splitting the sniff into several independent sniffs: https://github.com/squizlabs/PHP_CodeSniffer/issues/582

Basically we're already doing that now for WPCS and once the split off sniffs have been thoroughly tested, we should probably upstream all this.

**Comment by pento** [comment]
Possibly related, here's another "commas in array declarations" thing I've run into.

Given:

```php
$b = array('c'=>'d');
```

`WordPress.Arrays.ArrayDeclaration.NoComma` tries to add a comma before `'d'`, causing this:

```php
$b = array(
	'c' => ,'d'
);
```

It then gets stuck switching between `WordPress.Arrays.ArrayDeclaration.SpaceBeforeComma` and `WordPress.WhiteSpace.OperatorSpacing.NoSpaceAfter`, until it runs out of loops.

**Comment by jrfnl** [comment]
@pento Thanks for the additional input. That sounds like the same issue as #986. Correct ?
Both this issue + #986 should be fixed once I finished the new sniff.

**Comment by pento** [comment]
Ah, yes. The new one definitely looks like #986.
